File: footdev/_11주차/PGS_문자열_압축.java
package footdev._11주차;

import java.util.*;

class PGS_문자열_압축 {

    int min = Integer.MAX_VALUE;

    public int solution(String s) {

        //엣지 케이스 처리
        if (s.length() == 1) return 1;

        //로직
        int token = 1;
        int idx = 0;

        //자르는 단위가 길이의 반을 넘어서지 않아야함.
        while (token <= s.length() / 2) {
            idx = 0;
            Deque<String> q = new ArrayDeque<>();
            while (idx < s.length()) {
                if (idx + token > s.length()) {
                    q.add(s.substring(idx));
                    break;
                }
                String sub = s.substring(idx, idx + token);
                //큐가 비었으면 그냥 넣기
                if (q.isEmpty()) {
                    q.add(sub);
                }
                else {
                    //이전 토큰하고 같으면 뺀다음 카운트하기
                    if (q.peekLast().equals(sub)) {
                        //str은 이전 토큰
                        String str = q.pollLast();

                        //첫 연속적인 경우 처리
                        if (q.isEmpty() || !isNumberic(q.peekLast())) {
                            q.add("2");
                            q.add(str);
                        }
                        //2이상의 연속적인 경우 처리
                        else {
                            int cnt = Integer.parseInt(q.pollLast()) + 1;
                            q.add(String.valueOf(cnt));
                            q.add(str);
                        }
                    } else {
                        q.add(sub);
                    }
                }
                //idx를 토큰만큼 카운트
                idx += token;
            }

            StringBuilder sb = new StringBuilder();
            for (String str : q) {
                sb.append(str);
            }
            min = Math.min(min, sb.length());

            token++;
        }


        return min;
    }

    public boolean isNumberic(String str) {
        return str.chars().allMatch(Character::isDigit);
    }
}

File: daehwa_park/11주차/PGS_문자열_압축.py
def solution(s):
    min_len = float('inf')
    
    if len(s) == 1: # 문자열의 길이가 1이라면 무조건 1
        return 1
    
    for i in range(len(s) // 2): # 길이의 반보다 길게 나누면 어차피 압축이 안됨
        tmp_min = 0 # 임시 최솟값
        cnt = 0 # 반복 개수
        idx = 0 
        before = s[0:i+1] # 이전 문자열
        while idx + i < len(s):
            sub = s[idx:idx+i+1]
            if before == sub:
                cnt += 1
            else:
                if cnt > 1:
                    tmp_min += len(before) + len(str(cnt)) # 압축 횟수의 자릿수 만큼 더해줌
                else:
                    tmp_min += len(before)
                before = sub
                cnt = 1
            idx = idx + i + 1
            if idx + i >= len(s): # 자르고 남은 문자열
                tmp_min += len(s) - idx
                if cnt > 1:
                    tmp_min += len(before) + len(str(cnt))
                else:
                    tmp_min += len(before)
                break
        min_len = min(min_len, tmp_min)
        
    return min_len